tutorials/text_processing/WFST_Tutorial.ipynb

```diff
@@ -0,0 +1,8368 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 433,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "Xsgdu5TYx09_",
+    "outputId": "5812912f-883b-42e8-afbf-3ec4a0170345"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "249"
+      ]
+     },
+     "execution_count": 433,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "string_map_version = graph_tens_special\n",
+    "string_map_version.num_states()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "9jzn_U7s0Sit"
+   },
+   "source": [
+    "We see that their number of states (graph vertexes) are almost equal. Yet, if we use `pynini.optimize` - a method that calls a suite of WFST minimization algorithms: "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 434,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "7YtqhOY90iF0",
+    "outputId": "26f0f51b-b00d-4f5a-9b2f-330c9812666a"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "107"
+      ]
+     },
+     "execution_count": 434,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "constructed_version.optimize()\n",
+    "constructed_version.num_states()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 435,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "y93SqnOf0qa8",
+    "outputId": "74efcbfa-a272-4fc6-e36e-f1e31c6df221"
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "167"
+      ]
+     },
+     "execution_count": 435,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "string_map_version.optimize()\n",
+    "string_map_version.num_states()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "2cTdQj9L0xhl"
+   },
+   "source": [
+    "We see the latter possessing a significantly larger amount of graph vertices. \n",
+    "\n",
+    "So the decision will be dependent on your ITN needs, language, concern with efficiency, and design philosophy. Further, even decisions of language dialect will have an influence. \n",
+    "(e.g. Belgian, Canadian, and Swiss dialects of French will dispense with elements of the vigecimal system for the decimal schema.)\n",
+    "\n",
+    "**N.B.** *For reference: while `nemo_text_processing` grammars aim to minimize invalid productions, they assume input tokens are valid strings for a target language. (e.g. The mapping of \"quatre-vingt\" to `80` is permitted since it is not likely to occur in a valid French string.)*  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "V1djCnvY3CjW"
+   },
+   "source": [
+    "For more information on optimization algorithms for WFSTs, please see:\n",
+    "- K. Gorman and R. Sproat, *Finite-
```

Comment:
LOVE this! what do you think these notes at the top before you go into the examples? so readers are primed on what to expect and why you decided to build the grammars like this?

Reply:
Yeah, I want them to have a chance to stop and review if they get overwhelmed. As you say, it's a long tutorial so seems better to give stopping points throughout.